### DISCUSSION TITLE
In Chained Through Time pattern, the anchor shadow variable status does not match the planning variable in CH phase.

### DISCUSSION THREAD

**Discussion by kentbill** [TimefoldAI/timefold-solver#507] [General]
When using the Chained Through Time pattern, is there a situation where **the planning entity's planning variable is not empty, but its AnchorShadowVariable is empty?**

For example, in the TaskAssigning example(Just using this example to illustrate, not reproducible), a Task's planning variable - previousTaskOrEmployee is an Employee object, but its AnchorShadowVariable is empty (theoretically, at this point, the fields  previousTaskOrEmployee and employee should point to the same Employee object)

**The above situation only occurs in the Construction Heuristic phase.**
Thanks!

**Comment by kentbill** [comment]
this situation occurs in the afterVariableChanged method of VariableListener, that is, the second parameter(which is a planning entity object), its field previousTaskOrEmployee is not null but its anchor shadow variable - employee is null.

**Reply by Christopher-Chianelli** [reply]
Does your `@ShadowVariable` declare the `@AnchorShadowVariable` as a dependency?

i.e. is your code

```java
    @ShadowVariable(
            variableListenerClass = MyVariableListener.class,
            sourceVariableName = "previousTaskOrEmployee")
    @ShadowVariable(
            variableListenerClass = MyVariableListener.class,
            sourceVariableName = "anchor")
    MyObject shadowVariable;
```

or


```java
    @ShadowVariable(
            variableListenerClass = MyVariableListener.class,
            sourceVariableName = "previousTaskOrEmployee")
    MyObject shadowVariable;
```

If its the former, the anchor shadow variable should be defined when `MyVariableListener` is called.
If its the latter, Timefold is free to order the variable listeners like this: ["previousTaskOrEmployee", "shadowVariable", "anchor"], meaning when "shadowVariable" VariableListener is called, "anchor" will not be up-to-date when "previousTaskOrEmployee" will be up-to-date

**Reply by kentbill** [reply]
Thank you, Christopher!

Oh, it's amazing! After adding sourceVariableName="anchor" to the custom shadow variable - startTime, the parameter - task in method MyVariableListener. afterVariableChanged does not get such an inconsistency between the previous and anchor anymore.

If the cause is "Timefold is free to order the variable listeners," but in the Quickstarts example (such as call-center), for the shadow variable -  only previousCallOrAgent as the source variable of the custom shadow variable - **estimatedWaiting**, why is the situation - "previousCallOrAgent is not null but agent(anchor) is null" doesn't appear? Is there any other way to ensure the order of variable listeners?

**Reply by Christopher-Chianelli** [reply]
For the call-center example, its [the response time variable listener](https://github.com/TimefoldAI/timefold-quickstarts/blob/stable/use-cases/call-center/src/main/java/org/acme/callcenter/solver/ResponseTimeUpdatingVariableListener.java) does not ever use the agent shadow variable, so it does not need to list the agent shadow variable as a dependency. The anchor shadow variable is used in constraints (namely, "No required skills are missing" via `Call.getMissingSkillCount()`).

**Reply by kentbill** [reply]
Got it, thank you again!
